---
title: ""
output: 
  BiocStyle::html_document: 
    fig_width: 10
    fig_height: 8
---


```{r}
devtools::load_all("/Users/ricard/biofam/BioFAMtools", quiet=T)
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(purrr))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(reticulate))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(corrplot))
# library(BioFAMtools)
```

<!-- Define I/O and options -->
```{r}
source("/Users/ricard/gastrulation10x/mofa/load_settings.R")
```

<!-- Load sample metadata -->
```{r}
sample_metadata <- fread(io$sample.metadata) %>%
  .[,c("sample","stage","lineage10x","lineage10x_2","batch")]
```

```{r}
# sample_metadata[,embryonic:=lineage10x_2%in%c("Epiblast", "Primitive Streak", "Endoderm", "Ectoderm", "Mesoderm", "Blood", "PGC")]
```

```{r}
# asd <- read.table("/Users/ricard/data/gastrulation10x/mofa/data/E7.0/E7.0_14.txt.gz")
# asd <- read.table("/hps/nobackup2/research/stegle/users/ricard/gastrulation10x/mofa/data/E7.0/E7.0_14.txt.gz")
# 
# dim(asd)
# all(rownames(asd) %in% sample_metadata[sample=="14",cell])
```

<!-- Load model -->
```{r}
file <- paste0(io$basedir,"/mofa/hdf5/E7.0.hdf5")
model <- load_model(file)
```

<!-- Filter sample metadata -->
```{r}
samples <- as.character(unname(unlist(samples_names(model))))
sample_metadata_filt <- sample_metadata %>% setkey(sample) %>% .[samples]
stopifnot(all(samples==sample_metadata_filt$sample))
```

<!-- Rename groups -->
```{r}
# opts$groups_names <- c(
#   "atac.only" = "scATAC-seq only",
#   "coassay" = "sci-CAR (coassay)",
#   "rna.only" = "scRNA-seq only"
# )
# groups_names(model) <- stringr::str_replace_all(groups_names(model), opts$groups_names)
```

<!-- Subset factors -->
```{r}
r2 <- model@cache$variance_explained$r2_per_factor
factors <- sapply(r2, function(x) x[,"RNA"]>0.001)
model <- subset_factors(model, which(apply(factors,1,sum) >= 1))
factors_names(model) <- paste("Factor",1:get_dimensions(model)[["K"]], sep="")
```


<!-- Plot variance explained -->
```{r}
plot_variance_explained(model, x="group", y="factor")
```

<!-- Plot factors correlation -->
```{r}
plot_factor_cor(model)
```

<!-- Plot factors -->
```{r}
p <- plot_factors(
  model,
  factors = c(9),
  # color_by = "group",
  color_by = as.character(sample_metadata_filt$lineage10x),
  add_dots=T, add_violin=F, dodge=F
)

print(p)
# pdf(paste0(io$outdir,"/GOenrich_nonCGIprom_nondiff.pdf"), width=10, height=4, useDingbats = F)
# print(p)
# dev.off()
```


<!-- Plot weights -->
```{r}
view <- "RNA"
# view <- "ATAC_FALSE"

p <- plot_weights(
  model, 
  view=view, 
  factor=8, 
  nfeatures=10, 
  scale=F, abs=F
)
print(p)
```

<!-- Plot heatmaps -->
```{r}
model_i <- impute(model)
```

```{r}
anno_df <- sample_metadata_filt[,c("sample","group")] %>% 
  as.data.frame() %>% tibble::column_to_rownames("sample")
```

RNA
```{r}
pdf(paste0(io$pdfdir,"/Factor3_heatmap_rna_coassay.pdf"), useDingbats = F, onefile = F, width=6, height=3)

plot_data_heatmap(
  model3,
  view="RNA expression", 
  # groups = c("scRNA-seq only"),
  groups = c("sci-CAR (coassay)"),
  factor=3, 
  features=15,
  annotation_col=anno_df, annotation_colors=list("group"=opts$colors),
  color=colorRampPalette(brewer.pal(n = 9, name="YlGn"))(100),
  legend = T, annotation_legend=F, annotation_names_col=F,
  treeheight_row = 0, treeheight_col = 0,
  cluster_rows = T, cluster_cols=F,
  show_rownames = T, show_colnames=F
)

dev.off()
```




<!-- Gene set enrichment analysis -->

Load MsigDb gene set
```{r go_enrichment, echo=FALSE}
# I/O MSigDB
io$msigFile <- "/Users/ricard/data/MSigDB/v6.0/homo_sapiens/C5/bp_binary_matrix.rds"
# io$msigFile <- "/Users/ricard/data/MSigDB/v6.0/homo_sapiens/C2/binary_matrix.rds"

# Load MSigDB files
feature.sets <- readRDS(io$msigFile)
```

<!-- Rename features -->
```{r}
genes <- fread("/Users/ricard/data/ensembl/human/v87/BioMart/mRNA/Hsapiens_genes_BioMart.87.txt") %>%
  # .[symbol%in%features_names(model)[["RNA"]]] %>%
  .[ens_id%in%colnames(feature.sets)]
foo <- genes[,symbol]; names(foo) <- genes$ens_id

# new_names <- stringr::str_replace_all(colnames(feature.sets), foo)
new_names <- foo[colnames(feature.sets)]
new_names <- new_names[which(!is.na(new_names))]

feature.sets <- feature.sets[,names(new_names)]
colnames(feature.sets) <- new_names
```

```{r}
# BioFAMtools::features_names(model)[["RNA"]] <- toupper(BioFAMtools::features_names(model)[["RNA"]])
fsea.out <- FSEA(
  model,
  view = "RNA",
  feature.sets = feature.sets,
  statistical.test = "cor.adj.parametric",
  alpha = 0.01
)
```

```{r}
View(fsea.out$pval.adj)
```

```{r}
lineplot_FSEA(fsea.out, factor = 5, max.pathways = 8)
```


```{r}
library(umap)

algorithms <- c("umap")
views <- views_names(model)

umap.defaults$n_neighbors <- 20
umap.defaults$min_dist <- 0.7

samples <- as.character(unname(unlist(samples_names(model))))
```

<!-- Non-linear dimensionality reduction from MOFA factors -->
```{r}
# Fetch factors
Z <- get_factors(model) %>% do.call("rbind",.)

# Scale Z by the variance explained
# r2 <- model@cache$variance_explained$r2_per_factor
# Z.scaled <- Z %*% diag(r2*100)
Z.scaled <- Z

for (algorithm in algorithms) {
  
  set.seed(1)
  if (algorithm=="tsne") {
    tsne <- Rtsne(Z.scaled, check_duplicates=FALSE, pca=FALSE, theta=0.5, dims=2)
    Z.out <- tsne$Y
  } else if (algorithm=="umap") {
    umap.out <- umap(Z.scaled, config = umap.defaults)
    Z.out <- umap.out$layout
  }
  
  to.plot <- Z.out %>% as.data.table %>% .[,sample:=rownames(Z)] %>%
      merge(sample_metadata_filt, by="sample")

  p <- ggplot(to.plot, aes(x=V1, y=V2, color=`lineage10x`)) +
    geom_point(alpha=0.7, size=1) +
    scale_color_manual(values=opts$colors) +
    guides(colour = guide_legend(override.aes = list(size=3))) +
    labs(x="", y="") +
    theme_classic()
  
  # pdf(sprintf("%s/MOFA_%s_celltype.pdf",io$pdfdir,algorithm), width=6, height=5.5, useDingbats = F)
  print(p)
  # dev.off()
  
  # p2 <- ggplot(to.plot, aes(x=V1, y=V2, color=`sample`)) +
  #   geom_point(alpha=0.7, size=1.5) +
  #   theme_pub()
  # 
  # pdf(sprintf("%s/%s_MOFA_sample.pdf",io$pdfdir,algorithm), width=6, height=4.5, useDingbats = F)
  # print(p2)
  # dev.off()
}
```